### PR TITLE
[Artifacts] Add UID notation to artifact URI

### PR DESCRIPTION
**Pull request by yaelgen**
resolves: [ML-7169](https://iguazio.atlassian.net/browse/ML-7169) 

This PR modifies the artifact URI structure to include an optional `^uid` notation, enabling full support for querying artifacts by their unique identifier. 
The new format is: `{project}/{key}#{iteration}:{tag}@{tree}^{uid}`, it ensures consistency across MLRun objects while maintaining backward compatibility.  
Additionally, the `uid` is now included as part of the artifact model object.
Updates include enhancements to `parse_artifact_uri` and `generate_artifact_uri`, support for UID-based queries, and comprehensive tests to validate the changes.

For BC, as runs store a list of artifact URIs, older clients (pre-1.8.0) may attempt to access new URIs containing the `^uid` notation. Since these clients do not recognize the `^uid `field and may misinterpret it as part of the `tree`, the server has been updated to parse and handle the `^uid` field when reading artifacts.

[ML-7169]: https://iguazio.atlassian.net/browse/ML-7169?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mlrun/artifacts/base.py [hunk 1]
```diff
@@ -379,6 +379,7 @@ def get_store_url(self, with_tag=True, project=None, with_tree=True):
             iter=self.metadata.iter,
             tree=tree,
             tag=tag,
+            uid=self.metadata.uid,
```
Comment (TomerShor) [inline mlrun/artifacts/base.py]: Why not to use your new property
```suggestion
            uid=self.uid,
```

File: mlrun/artifacts/manager.py [hunk 2]
```diff
@@ -315,7 +315,13 @@ def log_artifact(
             item.upload(artifact_path=artifact_path)
 
         if db_key:
-            self._log_to_db(db_key, project, producer.inputs, item)
+            artifact_item = self._log_to_db(db_key, project, producer.inputs, item)
+            if artifact_item:
+                artifact_uid = artifact_item.get("metadata", {}).get("uid")
+                if artifact_uid is not None:
+                    item.uid = artifact_uid
+                    self.artifact_uris[key] = item.uri
```
Comment (TomerShor) [inline mlrun/artifacts/manager.py]: Why not return the `uid` from `_log_to_db` instead of the full item, since it's only used here.

File: mlrun/artifacts/manager.py [hunk 3]
```diff
@@ -327,20 +333,22 @@ def update_artifact(self, producer, item: Artifact):
...
+        self, key, project, sources, item, tag=None
+    ) -> typing.Optional[dict[str, str]]:
         """
         log artifact to db
         :param key: Identifying key of the artifact.
         :param project: Project that the artifact belongs to.
-        :param sources: List of artifact sources ( Mainly passed from the producer.items ).
+        :param sources: List of artifact sources ( Mainly passed from the `producer.items` ).
         :param item: The actual artifact to store.
         :param tag: The name of the Tag of the artifact.
```
Comment (TomerShor) [inline mlrun/artifacts/manager.py:341]: add return value to docstring

File: mlrun/db/httpdb.py [hunk 4]
```diff
@@ -995,7 +995,7 @@ def store_artifact(
         tag=None,
         project="",
         tree=None,
-    ):
+    ) -> dict[str, str]:
```
Comment (TomerShor) [inline mlrun/db/httpdb.py:999]: add return value to docstring

File: server/py/services/api/api/endpoints/artifacts_v2.py [hunk 5]
```diff
@@ -268,6 +268,12 @@ async def get_artifact(
             auth_info,
         )
     )
+
+    # Older clients (pre-1.8.0) do not support parsing "uid" and treat "tree^uid" as a single "tree" value.
+    # To ensure compatibility, we split "tree" here to extract "uid" if it exists.
+    if tree and "^" in tree:
+        tree, uid = tree.split("^", 1)
```
Comment (TomerShor) [inline server/py/services/api/api/endpoints/artifacts_v2.py]: 1. Note that `uid` is deprecated and we use `object_uid`.
2. What about the edge case where `"^" in tree and object_uid is not None`?

File: tests/utils/test_helpers.py [hunk 6]
```diff
@@ -369,6 +370,125 @@ def test_validate_artifact_name(artifact_name, expected):
...
+):
+    result = parse_artifact_uri(uri, project)
+    assert result == (
+        expected_project,
+        expected_key,
+        expected_iteration,
+        expected_tag,
+        expected_tree,
+        expected_uid,
+    ), f"Failed for URI: {uri}"
```
Comment (TomerShor) [inline tests/utils/test_helpers.py]:
```suggestion
    ), f"Failed to parse artifact URI: {uri}"
```

**Review by TomerShor — CHANGES_REQUESTED** [review]
Nice! have some suggestions and questions

File: server/py/services/api/api/endpoints/artifacts_v2.py (hunk 5 above)
Comment (liranbg) [reply server/py/services/api/api/endpoints/artifacts_v2.py]: as for (2), id omit the ^ from the tree if uid is none

File: mlrun/artifacts/manager.py [hunk 7]
```diff
@@ -315,7 +315,11 @@ def log_artifact(
             item.upload(artifact_path=artifact_path)
 
         if db_key:
-            self._log_to_db(db_key, project, producer.inputs, item)
+            artifact_uid = self._log_to_db(db_key, project, producer.inputs, item)
+            if artifact_uid is not None:
+                item.uid = artifact_uid
+                self.artifact_uris[key] = item.uri
```
Comment (moranbental) [inline mlrun/artifacts/manager.py]: Why are you updating the `artifact_uris` again? It has already been done in line 309
Is it because the URI now includes the UID? If so, add a comment

Or maybe it would be better to move line 309 to line 322 (after logging to the DB), regardless of whether there is a db_key or not

File: mlrun/projects/project.py [hunk 8]
```diff
@@ -3900,18 +3900,19 @@ def deploy_function(
...
 
-    def get_artifact(self, key, tag=None, iter=None, tree=None):
+    def get_artifact(self, key, tag=None, iter=None, tree=None, uid=None):
         """Return an artifact object
 
         :param key: artifact key
         :param tag: version tag
         :param iter: iteration number (for hyper-param tasks)
         :param tree: the producer id (tree)
+        :param uid: the artifact uid
```
Comment (moranbental) [inline mlrun/projects/project.py]:
```suggestion
        :param key: Artifact key
        :param tag: Version tag
        :param iter: Iteration number (for hyper-param tasks)
        :param tree: The producer id (tree)
        :param uid: The artifact uid
```

File: mlrun/projects/project.py [hunk 9]
```diff
@@ -3900,18 +3900,19 @@ def deploy_function(
             mock=mock,
         )
 
-    def get_artifact(self, key, tag=None, iter=None, tree=None):
+    def get_artifact(self, key, tag=None, iter=None, tree=None, uid=None):
```
Comment (moranbental) [inline mlrun/projects/project.py]:
```suggestion
    def get_artifact(self, key, tag=None, iter=None, tree=None, uid=None) -> typing.Optional[Artifact]:
```

File: mlrun/utils/helpers.py [hunk 10]
```diff
@@ -671,7 +671,7 @@ def dict_to_json(struct):
 def parse_artifact_uri(uri, default_project=""):
     """
     Parse artifact URI into project, key, tag, iter, tree
```
Comment (moranbental) [inline mlrun/utils/helpers.py]:
```suggestion
    Parse artifact URI into project, key, tag, iter, tree, uid
```

File: mlrun/utils/helpers.py [hunk 11]
```diff
@@ -719,14 +721,16 @@ def generate_object_uri(project, name, tag=None, hash_key=None):
     return uri
 
 
-def generate_artifact_uri(project, key, tag=None, iter=None, tree=None):
+def generate_artifact_uri(project, key, tag=None, iter=None, tree=None, uid=None):
```
Comment (moranbental) [inline mlrun/utils/helpers.py]:
```suggestion
def generate_artifact_uri(project, key, tag=None, iter=None, tree=None, uid=None) -> str:
```

File: server/py/services/api/crud/runs.py [hunk 12]
```diff
@@ -559,7 +559,7 @@ def _get_artifacts_from_uris(
         key_tag_iteration_pairs = []
         for _, uri in artifact_uris.items():
             _, uri = mlrun.datastore.parse_store_uri(uri)
-            project, key, iteration, tag, artifact_producer_id = (
+            project, key, iteration, tag, artifact_producer_id, uid = (
```
Comment (moranbental) [inline server/py/services/api/crud/runs.py:562]: add uid to the warning in line 566. 
I also think we should add it here: `key_tag_iteration_uid_pairs.append((key, tag, iteration, uid))` (line 576), and send it to `list_artifacts_for_producer_id`. This will help find unique artifacts

**Review by moranbental — CHANGES_REQUESTED** [review]
Looks good! Adding some comments. 
Also, `parse_artifact_uri` is called from here: https://github.com/mlrun/mlrun/blob/d3b3101a1b682cfed5ab68ca95dfb492e44c1146/server/py/services/api/crud/model_monitoring/model_endpoints.py#L227

File: server/py/framework/db/base.py [hunk 13]
```diff
@@ -255,7 +255,7 @@ def list_artifacts_for_producer_id(
         session,
         producer_id: str,
         project: str,
-        key_tag_iteration_pairs: list[tuple] = "",
+        artifact_identifiers: list[tuple] = "",
```
Comment (TomerShor) [inline server/py/framework/db/base.py:258]: you need to change this in the implementing class too (`SQLDB`)

File: server/py/framework/db/sqldb/db.py [hunk 14]
```diff
@@ -1590,7 +1590,7 @@ def _find_artifacts_for_producer_id(
         key_tag_iteration_pairs: list[tuple] = "",
```
Comment (TomerShor) [inline server/py/framework/db/sqldb/db.py]:
```suggestion
        artifact_identifiers: list[tuple] = "",
```

**Review by TomerShor — CHANGES_REQUESTED** [review]
Minor

File: server/py/framework/db/sqldb/db.py [hunk 15]
```diff
@@ -1587,14 +1587,14 @@ def _find_artifacts_for_producer_id(
...
+        artifact_identifiers: list[tuple] = "",
     ) -> list[tuple[ArtifactV2, str]]:
         """
-        Find a producer's artifacts matching the given (key, tag, iteration) tuples.
+        Find a producer's artifacts matching the given (key, tag, iteration, uid) tuples.
         :param session:                 DB session
         :param producer_id:             The artifact producer ID to filter by
         :param project:                 Project name to filter by
-        :param key_tag_iteration_pairs: List of tuples of (key, tag, iteration)
+        :param artifact_identifiers: List of tuples of (key, tag, iteration)
```
Comment (TomerShor) [inline server/py/framework/db/sqldb/db.py]:
```suggestion
        :param artifact_identifiers: List of tuples of (key, tag, iteration, uid)
```

**Review by TomerShor — CHANGES_REQUESTED** [review]

File: server/py/framework/db/sqldb/db.py [hunk 16]
```diff
@@ -1606,14 +1606,18 @@ def _find_artifacts_for_producer_id(
...
         tuples_filter = []
-        for key, tag, iteration in key_tag_iteration_pairs:
+        for key, tag, iteration, uid in artifact_identifiers:
             iteration = iteration or 0
             tag = tag or "latest"
-            tuples_filter.append(
+            base_filter = (
                 (ArtifactV2.key == key)
                 & (ArtifactV2.Tag.name == tag)
                 & (ArtifactV2.iteration == iteration)
```
Comment (moranbental) [inline server/py/framework/db/sqldb/db.py:1615]: Tag and iteration could also be None

**Review by moranbental — CHANGES_REQUESTED** [review]
minor

File: server/py/framework/db/sqldb/db.py (hunk 16 above)
Comment (liranbg) [reply server/py/framework/db/sqldb/db.py:1615]: 2 lines - they have defaults

File: server/py/services/api/api/endpoints/artifacts_v2.py [hunk 17]
```diff
@@ -268,6 +268,19 @@ async def get_artifact(
...
+
+    # Older clients (pre-1.8.0) do not support parsing "uid" and treat "tree^uid" as a single "tree" value.
+    # To ensure compatibility, we split "tree" here to extract "uid" if it exists.
+    if tree and "^" in tree:
+        tree, uri_object_uid = tree.split("^", 1)
+        if object_uid and object_uid != uri_object_uid:
+            mlrun.utils.logger.warning(
+                f"Conflicting UIDs detected: Provided object_uid='{object_uid}' "
+                f"does not match the UID extracted from the URI='{uri_object_uid}'."
+            )
```
Comment (liranbg) [inline server/py/services/api/api/endpoints/artifacts_v2.py]: note this is server-side log line, thus suggested to format it in a structural way
```suggestion
            mlrun.utils.logger.warning(
                "Conflicting UIDs detected", object_uid=object_uid, extracted_object_uid=uri_object_uid
            )
```

File: server/py/framework/db/sqldb/db.py (hunk 16 above)
Comment (yaelgen) [reply server/py/framework/db/sqldb/db.py:1615]: it's defaults to 0 and 'latest'

**Review by moranbental — APPROVED** [review]
LGTM

**Review by TomerShor — APPROVED** [review]
Nice 👌

**Review by liranbg — APPROVED** [review]
